staging/src/k8s.io/apiserver/pkg/endpoints/openapi/cacher_test.go

```diff
@@ -0,0 +1,321 @@
+		t.Errorf("Expected SchemaCacher to download a and parse a new spec but got %v and %v", firstGet, thirdGet)
+	}
+}
+
+// TestNewSchemaCacher is a full test of the schema cacher, it creates a new cacher that points
+// at fake openapi handler and Tests that it can get a proto format spec and parse it and look up a single resource
+func TestNewSchemaCacher(t *testing.T) {
+	fooPb, err := getReturnedPb(fooSwaggerJSON)
+	if err != nil {
+		t.Errorf("Error generating protobuf: %v\n", err)
+	}
+	handler := &fakeHandler{
+		swaggerPb: fooPb,
+		etag:      "firstEtag",
+	}
+	schemaCacher := NewSchemaCacher(handler, "/openapi/v2")
+
+	// We should only be able to look up Foo right now because Bar is not in the schema
+	resources, err := schemaCacher.Get()
+	if err != nil {
+		t.Errorf("Unexpected error getting schema: %v", err)
+	}
+	if err := expectLookupSuccess(resources, fooGVK); err != nil {
+		t.Error(err)
+	}
+}
+
+type fakeHandler struct {
+	swaggerPb []byte
+	etag      string
+}
+
+var _ http.Handler = &brokenHandler{}
+
+// ServeHTTP implements http.Handler
+func (f *fakeHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if len(f.etag) > 0 {
+		w.Header().Add("Etag", f.etag)
+	}
+	ifNoneMatches := r.Header["If-None-Match"]
+	for _, match := range ifNoneMatches {
+		if match == f.etag {
+			w.WriteHeader(http.StatusNotModified)
+			return
+		}
+	}
+
+	accept := r.Header.Get("Accept")
+	switch accept {
+	case "application/com.github.proto-openapi.spec.v2@v1.0+protobuf":
+		w.Write(f.swaggerPb)
+		w.Header().Set("Etag", f.etag)
+	default:
+		w.WriteHeader(http.StatusNotFound)
+		fmt.Fprintf(w, "Reqested format not implemented: %v\n", accept)
+	}
+}
+
+type brokenHandler struct{}
+
+var _ http.Handler = &brokenHandler{}
+
+// ServeHTTP implements http.Handler
+func (_ *brokenHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusNotFound)
+}
+
+type fakeDownloader map[string]downloaderResults
+type downloaderResults struct {
+	schemaBytes []byte
+	newEtag     string
+	httpStatus  int
+	err         error
+}
+
+var _ Downloader = make(fakeDownloader, 0)
+
+// DownloadSchema implements Downloader
+func (f fakeDownloader) DownloadSchema(lastEtag string) (schemaBytes []byte, newEtag string, httpStatus int, err error) {
+	results, ok := f[lastEtag]
+	if !ok {
+		// Return value not specified, coder error.
+		panic(nil)
+	}
+	return results.schemaBytes, results.newEtag, results.httpStatus, results.err
+}
+
+type fakeParser struct {
+	counter int
+}
+
+var _ Parser = &fakeParser{}
+
+// Parse implements Parser
+func (f *fakeParser) Parse(_ []byte) (Resources, error) {
+	f.counter = f.counter + 1
+	r := fakeResources(f.counter)
+	return r, nil
+}
+
+type fakeResources int
+
+var _ Resources = fakeResources(0)
+
+// LookupResource implements Resources
+func (_ fakeResources) LookupResource(gvk schema.GroupVersionKind) openapiproto.Schema {
+	panic(nil)
+}
+
+func getReturnedPb(swaggerJSON []byte) ([]byte, error) {
```

Comment:
jsonToPB?

style nit: acronyms should be same case, so "PB" instead of "Pb".

Reply:
I think I got this Pb idea from [kube-openapi](https://github.com/kubernetes/kube-openapi/blob/master/pkg/handler/handler.go#L65)

Reply:
The style thing is a losing battle :)